text/0101-revamping-the-sdk-performance-api.md

```diff
@@ -0,0 +1,268 @@
+
+There are two top level methods we'll be introducing to achieve this: `Sentry.startActiveSpan` and `Sentry.startSpan`. `Sentry.startActiveSpan` will take a callback and start/stop a span automatically. In addition, it'll also set the span on the current scope. Under the hood, the SDK will create a transaction or span based on if there is already an existing span on the scope.
+
+```ts
+namespace Sentry {
+  declare function startActiveSpan<T>(
+    spanContext: SpanContext,
+    callback: (span: Span) => T
+  ): T;
+}
+
+// span that is created is provided to callback in case additional
+// attributes have to be added.
+// ideally callback can be async/sync
+const returnVal = Sentry.startActiveSpan(
+  { name: "expensiveCalc" },
+  (span: Span) => expensiveCalc()
+);
+
+// If the SDK needs async/sync typed different we can expose this
+// declare function startActiveSpanAsync<T>(
+//   spanContext: SpanContext,
+//   callback: (span: Span) => Promise<T>,
+// ): Promise<T>;
+```
+
+In the ideal case, `startActiveSpan` should generally follow this code path.
+
+1. Get the active span from the current scope
+2. If the active span is defined, create a child span of that active span based on the provided `spanContext`, otherwise create a transaction based on the provided `spanContext`.
+3. Run the provided callback
+4. Finish the child span/transaction created in step 2
+5. Remove the child span/transaction from the current scope and if it exists, set the previous active span as the active span in the current scope.
+
+If the provided callback throws an exception, the span/transaction created in step 2 should be marked as errored. This error should not be swallowed by `startActiveSpan`.
+
+`Sentry.startSpan` will create a span, but not set it as the active span in the current scope. 90% of our documentation and code examples will be referencing `Sentry.startActiveSpan` - `Sentry.startSpan` is just there for edge case users of the API.
+
+```ts
+namespace Sentry {
+  declare function startSpan(spanContext: SpanContext): Span;
+}
+
+// does not get put on scope
+const span = Sentry.startSpan({ name: "expensiveCalc" });
+
+expensiveCalc();
+
+span.finish();
+```
+
+The only methods that all SDKs are required to implement are `Sentry.startActiveSpan` and `Sentry.startSpan`. For languages that need it, they add an additional method for async callbacks: `Sentry.startActiveSpanAsync`. Other languages can also attach a suffix to the methods to indicate that the spans are being started from different sources, but these are language/framework/sdk dependent.
+
+For example with go:
+
+```go
+sentry.StartSpanFromContext(ctx, spanCtx)
+```
+
+Or when continuing from headers in javascript:
+
+```js
+Sentry.startSpanFromHeaders(spanCtx, headers);
+```
+
+Since we want to discourage accessing the transaction object directly, the `Sentry.setMeasurement` top level method will also be introduced. This will set a custom performance metric if a transaction exists.
```

Comment:
Would be the case that you set a `measurement` to a specific span, which is a span in our data model, and we don't have the `measurements` field in the span data model, only in the Transaction data model, how would we make this difference? are we planning to introduce a new data model and give up on Transaction and Span?